### PR TITLE
Add CohereAI embeddings

### PR DESCRIPTION
**Pull request by ravi03071991**
# Description

PR for CohereAI Embeddings based on the latest [release](https://txt.cohere.com/introducing-embed-v3/)

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama_index/embeddings/cohereai.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    "embed-multilingual-light-v3.0",
+]
+
+DEFAULT_INPUT_TYPES = [
+    "search_query",
+    "search_document",
+    "classification",
+    "clustering",
+]
+DEFAULT_TRUNCATE_OPTIONS = ["START", "END", "NONE"]
```
Comment (nerdai) [inline llama_index/embeddings/cohereai.py:21]: I wonder if we can follow a similar pattern to how we handle this type of defining "modes" and "model" as we did in `OpenAIEmbeddings`. That is, using Enum and Dicts.

File: llama_index/embeddings/cohereai.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                    f"input_type must be one of {DEFAULT_INPUT_TYPES} for model '{model_name}'"
+                )
+        else:
+            if input_type is not None:
+                raise ValueError(
+                    f"input_type should not be provided for model '{model_name}'"
+                )
+
+        if truncate not in DEFAULT_TRUNCATE_OPTIONS:
+            raise ValueError(f"truncate must be one of {DEFAULT_TRUNCATE_OPTIONS}")
```
Comment (nerdai) [inline llama_index/embeddings/cohereai.py:64]: if we are able to use Enum/Dict pattern, then I think this validation would just be checking if keys exist in our dictionaries.

See for example [here](https://github.com/run-llama/llama_index/blob/f7fb2de6c5cbc54f682c538298bb2e64929f46d9/llama_index/embeddings/openai.py#L195-L198)

File: llama_index/embeddings/cohereai.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        )
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "CohereEmbedding"
+
+    def _embed(self, texts: List[str]) -> List[List[float]]:
+        """Embed sentences using Cohere."""
+        if self.input_type:
+            embeddings = self.cohere_client.embed(
```
Comment (nerdai) [inline llama_index/embeddings/cohereai.py:83]: very nit (and feel free to ignore): use `result` in favor of `embedding` so its not `embedding.embeddings` but rather `result.embeddings`

**Review by nerdai — APPROVED** [review]
Looks pretty good to me. I do wonder though if we can use a similar implementation to defining constants/defaults as we did in OpenAIEmbeddings for increased symmetry.

File: llama_index/embeddings/cohereai.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+DEFAULT_TRUNCATE_OPTIONS = ["START", "END", "NONE"]
+
+
+# Assuming BaseEmbedding is a Pydantic model and handles its own initializations
+class CohereEmbedding(BaseEmbedding):
+    """CohereEmbedding uses the Cohere API to generate embeddings for text."""
+
+    # Instance variables initialized via Pydantic's mechanism
+    cohere_client: Any = Field(description="CohereAI client")
+    model_name: str = Field(description="CohereAI model name")
```
[the remark this answers is not in the document]
Comment (ravi03071991) [reply llama_index/embeddings/cohereai.py:30]: But the cohere_client is needed I guess?

[the remark this answers is not in the document]
Comment (nerdai) [reply llama_index/embeddings/cohereai.py:30]: yeah it is. I guess you can say this comment is pretty nit 😃

**Review by nerdai — APPROVED** [review]
LGTM!

**Pull request merged**